rviz_rendering/src/rviz_rendering/grid.cpp

```diff
@@ -138,105 +139,117 @@ void Grid::setStyle(Style style)
+
+  createLines(addLineFunction);
+}
+
+uint32_t Grid::numberOfVerticalLines() const
+{
+  return (cell_count_ + 1) * (cell_count_ + 1) * height_count_;
+}
+
+void Grid::createLines(AddLineFunction addLine) const
+{
+  float plane_extent = (cell_length_ * static_cast<float>(cell_count_)) / 2;
+
+  for (uint32_t current_height = 0; current_height <= height_count_; ++current_height) {
+    createGridPlane(plane_extent, current_height, addLine);
   }
 
-  if (height_ > 0) {
-    for (uint32_t x = 0; x <= cell_count_; ++x) {
-      for (uint32_t z = 0; z <= cell_count_; ++z) {
-        float x_real = extent - x * cell_length_;
-        float z_real = extent - z * cell_length_;
-
-        float y_top = (height_ / 2.0f) * cell_length_;
-        float y_bottom = -y_top;
-
-//        if (style_ == Billboards)
-//        {
-//          billboard_line_->newLine();
-//
-//          billboard_line_->addPoint( Ogre::Vector3(x_real, y_bottom, z_real));
-//          billboard_line_->addPoint( Ogre::Vector3(x_real, y_top, z_real));
-//        }
-//        else
-//        {
-        manual_object_->position(x_real, y_bottom, z_real);
-        manual_object_->colour(color_);
-        manual_object_->position(x_real, y_top, z_real);
-        manual_object_->colour(color_);
-//        }
-      }
-    }
+  if (height_count_ > 0) {
+    createVerticalLinesBetweenPlanes(plane_extent, addLine);
   }
+}
+
+void Grid::createGridPlane(float extent, uint32_t height, AddLineFunction addLine) const
+{
+  float real_height = (height_count_ / 2.0f - static_cast<float>(height)) * cell_length_;
+  for (uint32_t i = 0; i <= cell_count_; i++) {
+    float inc = extent - ( i * cell_length_);
+
+    Ogre::Vector3 p1(inc, real_height, -extent);
+    Ogre::Vector3 p2(inc, real_height, extent);
+    Ogre::Vector3 p3(-extent, real_height, inc);
+    Ogre::Vector3 p4(extent, real_height, inc);
+
+    addLine(p1, p2);
+    addLine(p3, p4);
+  }
+}
 
-  if (style_ == Lines) {
-    manual_object_->end();
+void Grid::createVerticalLinesBetweenPlanes(float extent, AddLineFunction addLine) const
+{
+  for (uint32_t x = 0; x <= cell_count_; ++x) {
+    for (uint32_t z = 0; z <= cell_count_; ++z) {
+      float x_real = extent - x * cell_length_;
+      float z_real = extent - z * cell_length_;
+
+      float y_top = (height_count_ / 2.0f) * cell_length_;
+      float y_bottom = -y_top;
+
+      Ogre::Vector3 p1(x_real, y_bottom, z_real);
+      Ogre::Vector3 p2(x_real, y_top, z_real);
+
+      addLine(p1, p2);
+    }
   }
 }
 
+void Grid::addManualLine(const Ogre::Vector3 & p1, const Ogre::Vector3 & p2) const
+{
+  manual_object_->position(p1);
+  manual_object_->colour(color_);
+  manual_object_->position(p2);
+  manual_object_->colour(color_);
+}
+
+void Grid::addBillboardLine(const Ogre::Vector3 & p1, const Ogre::Vector3 & p2) const
+{
+  billboard_line_->addPoint(p1);
+  billboard_line_->addPoint(p2);
+  billboard_line_->newLine();
```

Comment:
This won't be possible without allowing that additional newLine() -> after the last points have been added, we'd need a check "if (not last line)" or "if (not first line)" .